### PR TITLE
Support array literal with scalar function

### LINKED ISSUE #8867: Calling make_array() on struct literal causes panic
### Describe the bug

Attempting to create a literal value which is an array of structs causes a panic.

### To Reproduce

```
DataFusion CLI v34.0.0
❯ set datafusion.execution.parquet.allow_single_file_parallelism=false;
0 rows in set. Query took 0.001 seconds.

❯ copy (values (make_array(struct('foo',1)))) to '/tmp/test.parquet';
thread 'main' panicked at /home/dev/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-common-34.0.0/src/scalar.rs:524:74:
called `Result::unwrap()` on an `Err` value: Internal("Unsupported data type in hasher: Struct([Field { name: \"c0\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: \"c1\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }])")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Expected behavior

Should not result in panic. If there is some issue with how I have attempted to construct the array of struct, we could add a more descriptive error message.

### Additional context

I came across this while writing tests related to #8551, but this issue does not appear related to parallel parquet writing specifically.

The line of code generating the error implies an internal constraint is violated if we reach it.

https://github.com/apache/arrow-datafusion/blob/1dcdcd431187178d736cdd3a6c004204aa2faa14/datafusion/common/src/hash_utils.rs#L369-L374

### LINKED ISSUE #8873: Support array literal with struct
### Describe the bug

array literal with struct occurs error, the reason is that we only consider Array function as valid type. It seems we should also accept Struct 

```rust
                    if fun == BuiltinScalarFunction::MakeArray {
                        values.push(value);
                    } else {
                        // BuiltinScalarFunction::Struct

                        return not_impl_err!(
                            "ScalarFunctions without MakeArray are not supported: {value}"
                        );
                    }
```

### To Reproduce
```
query error DataFusion error: This feature is not implemented: ScalarFunctions without MakeArray are not supported: struct\(Utf8\("x"\), Int64\(1\)\)
select [struct('x', 1)];
```
### Expected behavior

We can see make_array did the good job here.
```
query ?
select make_array(struct('x', 1));
----
[{c0: x, c1: 1}]
```

### Additional context

Bug found while looking into #8867

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #8873
Closes #8867

### CODE REVIEW

File: datafusion/sqllogictest/test_files/array.slt [hunk 1]
```diff
@@ -459,7 +459,7 @@ AS
 query error
 select [1, true, null]
 
-query error DataFusion error: This feature is not implemented: ScalarFunctions without MakeArray are not supported: now()
```
Comment (jackwener) [inline datafusion/sqllogictest/test_files/array.slt:462]: Can we add more to covert `Struct`?

**Review by alamb — APPROVED** [review]
looks good to me, thank you @jayzhan211 

Now that I am looking at this code, I wonder why array literals can't support functions 🤔  I bet they could given they are just arguments passed to a function now

**Comment by jayzhan211** [comment]
> looks good to me, thank you @jayzhan211
> 
> Now that I am looking at this code, I wonder why array literals can't support functions 🤔 I bet they could given they are just arguments passed to a function now

True

File: datafusion/sql/src/expr/value.rs [hunk 2]
```diff
@@ -135,38 +135,12 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                    values.push(value);
-                }
-                _ => {
-                    return internal_err!(
-                        "Arrays with elements other than literal are not supported: {value}"
-                    );
-                }
-            }
-        }
+        let values = elements
```
Comment (alamb) [inline datafusion/sql/src/expr/value.rs:138]: I love it -- less code and more functionality ❤️ !

**Review by alamb — APPROVED** [review]

**Comment by jayzhan211** [comment]
Forgot to convert back from draft, this is ready to merge!

**Review by jackwener — APPROVED** [review]

**Pull request merged**